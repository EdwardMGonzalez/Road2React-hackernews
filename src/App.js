import React, { Component } from "react";
import "./App.css";
import PropTypes from "prop-types";
import axios from "axios";

const PATH_BASE = "https://hn.algolia.com/api/v1";
const PATH_SEARCH = "/search_by_date";
const PARAM_SEARCH = "query=";
const PARAM_PAGE = "page=";

class App extends Component {
  __isMounted = false;

  constructor(props) {
    super(props);

    this.state = {
      result: null,
      searchTerm: "",
      isLoading: true,
      error: null
    };

    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
  }

  setSearchTopStories(result) {
    if (this.__isMounted) {
      this.setState({ result });
      this.setState( {isLoading:false} );
    }
  }

  componentDidMount() {
    this.__isMounted = true;
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    this.__isMounted = false;
  }

  onDismiss(id) {
    const isNotId = item => item.objectID !== id;
    const updatedHits = this.state.result.hits.filter(isNotId);
    this.setState({
      result: Object.assign({}, this.state.result, { hits: updatedHits })
    });
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
    event.preventDefault();
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    this.setState( {isLoading:true} );
    axios
      .get(
        `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`
      )
      .then(result => this.setSearchTopStories(result.data))
      .catch(error => this.setState({ error }));
  }

  render() {
    const { searchTerm, result, error, isLoading } = this.state;
    const page = (result && result.page) || 0;

    return (
      <div className="page">
        <h1>Keyword Search</h1>
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
          {error ? <p>Something went wrong.</p> : null}
          {isLoading ? <Loading /> : null }
          {result && !isLoading ? (
            <div>
              <Pagination
                page={result.page}
                maxPage={result.nbPages}
                searchTerm={searchTerm}
                goBack={() => this.fetchSearchTopStories(searchTerm, page - 1)}
                goForward={() => this.fetchSearchTopStories(searchTerm, page + 1)}
              />
              <Table
                list={result.hits}
                pattern={searchTerm}
                onDismiss={this.onDismiss}
              />
              <Pagination
                page={result.page}
                maxPage={result.nbPages}
                searchTerm={searchTerm}
                goBack={() => this.fetchSearchTopStories(searchTerm, page - 1)}
                goForward={() => this.fetchSearchTopStories(searchTerm, page + 1)}
              />
            </div>
          ) : null}
        </div>
      </div>
    );
  }
}
class Search extends Component {
  componentDidMount() {
      if(this.input) {
        this.input.focus();
      }
  }

  render() {
    const { onSubmit, onChange, value, children } = this.props;
    return (
      <form onSubmit={onSubmit}>
        <input
          type="text"
          onChange={onChange}
          value={value}
          ref={el => (this.input = el)}
        />
        <button type="submit">{children}</button>
      </form>
    );
  }
}

Search.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.string
};

const Table = ({ list, pattern, onDismiss }) => (
  <div className="table">
    {list.map(item => (
      <div
        key={item.objectID}
        className="table-row"
        id={"row-" + item.objectID}
      >
        <span className="title">
          <a href={item.story_url ? item.story_url : item.url}>
            {item.story_title ? item.story_title : item.title}
          </a>
        </span>
        <span className="author">{item.author}</span>
        <span className="num_comments">{item.num_comments}</span>
        <span className="points">{item.points}</span>
        <span className="button-inline">
          <Button
            onClick={() => onDismiss(item.objectID)}
            className="button-inline"
          >
            Dismiss
          </Button>
        </span>
      </div>
    ))}
  </div>
);

Table.propTypes = {
  list: PropTypes.array.isRequired,
  pattern: PropTypes.string,
  onDismiss: PropTypes.func.isRequired
};

const Button = ({ onClick, className = "", children }) => (
  <button onClick={onClick} className={className} type="button">
    {children}
  </button>
);

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.any
};

const Loading = () => (
  <div>Loading...</div>
);

const Pagination = ( {page = 0, maxPage, goBack, goForward} ) => (
  <div>
    <span>
      {page !== 0
        ?
        <Button  onClick={goBack} >
          &lt;
        </Button>
      : null }
      {page + 1} of {maxPage}
      { page !== maxPage - 1
        ?
        <Button  onClick={goForward} >
          &gt;
        </Button>
        : null }
    </span>
  </div>
);

export default App;

export { Button, Search, Table };
