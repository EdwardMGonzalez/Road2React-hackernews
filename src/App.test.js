import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import App, { Search, Button, Table } from "./App";

Enzyme.configure({ adapter: new Adapter() });

describe("App", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const searchProps = {
  value: "",
  onChange: () => {},
  onSubmit: () => {},
  children: ""
};

describe("Search", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Search {...searchProps}>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Search {...searchProps}>Potato</Search>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const buttonProps = {
  onClick: () => {},
  className: "button",
  children: "Check me out!"
};

describe("Button", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Button {...buttonProps}>Click Me!</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(
      <Button {...buttonProps}>Click Me!</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const storyTitle =
  "Algorithm recovers speech from a potato-chip bag filmed through glass (2014)";
const storyUrl =
  "http://news.mit.edu/2014/algorithm-recovers-speech-from-vibrations-0804";
const title = "Potato paradox";
const url = "https://en.wikipedia.org/wiki/Potato_paradox";
const tableProps = {
  list: [
    {
      created_at: "2015-07-15T21:37:09.000Z",
      title: title,
      url: url,
      author: "Panoramix",
      points: 559,
      story_text: null,
      story_title: null,
      story_url: null,
      objectID: "9894237"
    },
    {
      created_at: "2018-08-05T17:40:44.000Z",
      title: null,
      url: null,
      author: "MrJagil",
      points: 327,
      story_text: null,
      story_title: storyTitle,
      story_url: storyUrl,
      objectID: "17692447"
    }
  ],
  onDismiss: () => {}
};

describe("Table", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Table {...tableProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Table {...tableProps} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows two items in list", () => {
    const element = shallow(<Table {...tableProps} />);
    expect(element.find(".table-row").length).toBe(2);
  });

  it("uses objectID for row Ids", () => {
    const element = shallow(<Table {...tableProps} />);
    expect(element.find("#row-17692447").length).toBe(1);
    expect(element.find("#row-9894237").length).toBe(1);
  });

  it("shows title based on item.story_title and uses item.story_url", () => {
    const element = shallow(<Table {...tableProps} />);
    const row = element.find("#row-17692447");
    expect(
      row.containsMatchingElement(<a href={storyUrl}>{storyTitle}</a>)
    ).toEqual(true);
  });

  it("shows title based on item.title and uses item.url", () => {
    const element = shallow(<Table {...tableProps} />);
    const row = element.find("#row-9894237");
    expect(row.containsMatchingElement(<a href={url}>{title}</a>)).toEqual(
      true
    );
  });
});
