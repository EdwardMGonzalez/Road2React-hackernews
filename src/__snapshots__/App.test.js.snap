// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <h1>
    Keyword Search
  </h1>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className="button"
  onClick={[Function]}
  type="button"
>
  Click Me!
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    type="submit"
  >
    Potato
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
    id="9894237"
  >
    <span
      className="title"
    >
      <a
        href="https://en.wikipedia.org/wiki/Potato_paradox"
      >
        Potato paradox
      </a>
    </span>
    <span
      className="author"
    >
      Panoramix
    </span>
    <span
      className="num_comments"
    />
    <span
      className="points"
    >
      559
    </span>
    <span
      className="button-inline"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
    id="17692447"
  >
    <span
      className="title"
    >
      <a
        href="http://news.mit.edu/2014/algorithm-recovers-speech-from-vibrations-0804"
      >
        Algorithm recovers speech from a potato-chip bag filmed through glass (2014)
      </a>
    </span>
    <span
      className="author"
    >
      MrJagil
    </span>
    <span
      className="num_comments"
    />
    <span
      className="points"
    >
      327
    </span>
    <span
      className="button-inline"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
